examples/cfq/syntax-based/grammar.py

```diff
@@ -74,29 +77,98 @@ def generate_grammar(grammar_str: str):
     rules.append(current_rule)
   sub_rules = []
   for rule in rules:
-    sub_rules += generate_sub_rules(rule)
+    sub_rules += extract_sub_rules_from_rules(rule)
+  return sub_rules
+
+class TermType(Enum):
+  SYNTAX_TERM = 0
+  REGEX_TERM = 1
+  RULE_TERM = 2
+
+class Term:
+  """
+  Class describing a term (token on the right side of a rule). It contains a
+  type and a value. If the term is a syntax token, eg. "SELECT", the value
+  contains the string "SELECT", if the term is a regex term, it also contains a
+  string, for eg. "/[^\s]+/", and if the term is a rule head, the head is stored.
+  """
+
+  def __init__(self, term_type: TermType, value: str):
+    self.term_type = term_type
+    self.value = value
+
+  def __repr__(self):
+    return self.value
 
-  sub_rules_dict = {f'r{i}': sub_rules[i] for i in range(len(sub_rules))}
-  return sub_rules_dict
+  def __eq__(self, other):
+    """Overrides the default implementation"""
+    return self.term_type == other.term_type and self.value == other.value
 
-def generate_rules_by_head(sub_rules_dict: Dict):
-  result = collections.defaultdict(list)
-  for rule_name, (head, body) in sub_rules_dict.items():
-    result[head].append((rule_name, body))
-  return result
+class RuleBranch:
```

Comment:
I like that you added a docstring to class `Term`, could you do the same for this class?